File: dlr_sensors/src/fake_gps_sensor_class_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/NavSatFix.h>
#include <std_srvs/SetBool.h>
#include "dlr_sensors/FakeGPS.h"

namespace dlr{

class FakeGPSNode
{
    public: 
        // Class constructor
        FakeGPSNode() : nh_("~"), active_(false), activation_service_name_("activate")
        {
            pub_ = nh_.advertise<sensor_msgs::NavSatFix>("data", 10);
            activation_service_ = nh_.advertiseService(activation_service_name_, &FakeGPSNode::activationServiceCallback, this);
        };

        bool init()
        {
            return getROSParams();
        }

        bool updateData(const float& elapsed_time)
        {
            return gps_driver_.updateData(elapsed_time);
        }

        bool publishData()
        {
            if (active_)
            {
                if ((pub_.getNumSubscribers() == 0))
                {
                    ROS_WARN_STREAM_THROTTLE(10, "GPS active, but no node is subscribed to \"" << 
                                    nh_.resolveName("data") << "\" ROS topic.");
                    return false;
                }
                sensor_msgs::NavSatFix msg;
                msg.header.stamp = ros::Time().now();
                msg.header.frame_id = frame_id_;
                msg.status.status = sensor_msgs::NavSatStatus().STATUS_FIX;
                //msg.status.service = NavSatStatus.SERVICE_GPS
                msg.latitude = gps_driver_.getLatitude();
                msg.longitude = gps_driver_.getLongitude();
                msg.altitude = gps_driver_.getAltitude();
                
                msg.position_covariance[0] = 0;
                msg.position_covariance[4] = 0;
                msg.position_covariance[8] = 0;
                msg.position_covariance_type = sensor_msgs::NavSatFix().COVARIANCE_TYPE_DIAGONAL_KNOWN;

                ROS_INFO_STREAM("Actual position: [latitude: " << msg.latitude << 
                                        " | longitude: " << msg.longitude << "]");

                pub_.publish(msg);
            }
            else
            {
                ROS_WARN_STREAM_THROTTLE(10, "Fake GPS sensor node waiting for activation. Call \"" << 
                                        nh_.resolveName(activation_service_name_) << "\" service using the following command: " <<
                                        "rosservice call " << nh_.resolveName(activation_service_name_) << " \"data: true\"");
            }
            
            return true;
        };

    private:

        ros::NodeHandle nh_;
        ros::Publisher pub_;
        ros::ServiceServer activation_service_;
        std::string activation_service_name_;
        dlr::FakeGPS gps_driver_;

        bool active_;

        float position_covariance_;
        std::string frame_id_;

        bool activationServiceCallback(std_srvs::SetBool::Request& req, std_srvs::SetBool::Response& res)
        {
            if (req.data)
            {
                res.message = "GPS sensor activated.";
                res.success = true;
                ROS_INFO_STREAM(res.message);
                active_ = true;
            }
            else
            {
                res.message = "GPS sensor deactivated.";
                res.success = true;
                ROS_INFO_STREAM(res.message);
                active_ = false;
            }
            return true;
        }

        bool getROSParams()
        {
            // Get a ROS parameter, if does not exists initialization fails
            if (!nh_.getParam("frame_id", frame_id_))
            {
                ROS_ERROR_STREAM("No " << nh_.resolveName("frame_id") << " ROS param found.");
                return false;
            }
            
            // Get a ROS parameter, if does not exists assign a default value
            if (!nh_.param<float>("position_covariance", position_covariance_, 0.005f))
            {
                ROS_WARN_STREAM("No " << nh_.resolveName("position_covariance") << " ROS param found. Assigning default value:" << position_covariance_);
            }
            return true;
        }

}; // class FakeGPSNode

} // namespace dlr

int main(int argc, char** argv)
{
    // Initialize the ROS node 
    ros::init(argc, argv, "gps");

    // Declare GPS object node
    dlr::FakeGPSNode gps_node;

    // GPS node initialization
    if (!gps_node.init())
    {
        ROS_ERROR("Initialization error.");
        return -1;
    }

    // Define the loop rate
    float loop_rate_hz = 1.0;
    ros::Rate loop_rate(loop_rate_hz);

    ROS_INFO("GPS node started.");
    
    while (ros::ok())
    {
        gps_node.updateData(1.0/loop_rate_hz);
        gps_node.publishData();

        ros::spinOnce();
        loop_rate.sleep();
    }

    return 0;
}

File: dlr_sensors/scripts/fake_gps_sensor_class_node.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import NavSatFix, NavSatStatus
from std_srvs.srv import SetBool, SetBoolResponse

class FakeGPSNode:
    def __init__(self, activation_service_name="activate"):
        self.__active = False
        self.__activation_service_name = "~"+activation_service_name
        self.__pub = rospy.Publisher("~data", NavSatFix, queue_size=10)
        self.__activation_service = rospy.Service(self.__activation_service_name, SetBool, self.activation_service_callback)
        self.__latitude = 52.25444
        self.__longitude = 13.31476
        self.__altitude = 12.345
        self.__constant_speed = 0.001
        self.__frame_id = "gps_base_link"
        self.__position_covariance = 0.0

    def init(self):
        return self.get_ROS_params()

    def update_data(self, elapsed_time):
        self.__latitude += (elapsed_time * self.__constant_speed)
        self.__longitude -= (elapsed_time * self.__constant_speed)
        self.__altitude = self.__altitude
        return True

    def publish_data(self):
        if self.__active:
            if self.__pub.get_num_connections() == 0:
                rospy.logwarn_throttle(10, "GPS active, but no node is subscribed to \"" +
                                    rospy.resolve_name("data") + "\" ROS topic.")
                return False
            
            msg = NavSatFix()
            msg.header.stamp = rospy.Time().now()
            msg.header.frame_id = self.__frame_id
            msg.status.status = NavSatStatus().STATUS_FIX
            #msg.status.service = NavSatStatus().SERVICE_GPS
            msg.latitude = self.__latitude
            msg.longitude = self.__longitude
            msg.altitude = self.__altitude
                
            msg.position_covariance[0] = self.__position_covariance
            msg.position_covariance[4] = self.__position_covariance
            msg.position_covariance[8] = self.__position_covariance
            msg.position_covariance_type = NavSatFix().COVARIANCE_TYPE_DIAGONAL_KNOWN

            rospy.loginfo("Actual position: [latitude: " + str(msg.latitude) + 
                                    " | longitude: " + str(msg.longitude) + "]")

            self.__pub.publish(msg)
        else:
            rospy.logwarn_throttle(10, "Fake GPS sensor node waiting for activation. Call \"" +
                                        rospy.resolve_name(self.__activation_service_name) + "\" service using the following command: " +
                                        "rosservice call " + rospy.resolve_name(self.__activation_service_name) + " \"data: true\"")
            return False
        
        return True

    def activation_service_callback(self, req):
        res = SetBoolResponse()
        if req.data:
            res.message = "GPS sensor activated."
            res.success = True
            rospy.loginfo(res.message)
            self.__active = True
        else:
            res.message = "GPS sensor deactivated."
            res.success = True
            rospy.loginfo(res.message)
            self.__active = False
        return res

    def get_ROS_params(self):
        # Get a ROS parameter, if does not exists initialization fails
        try:
            self.__frame_id = rospy.get_param("~frame_id")
        except:
            rospy.logerr("No " + rospy.resolve_name("frame_id") + " ROS param found.")
            return False
        
        # Get a ROS parameter, if does not exists assign a default value
        try:
            self.__position_covariance = rospy.get_param("position_covariance", 0.005)
        except:
            rospy.logwarn("No " + rospy.resolve_name("position_covariance") + " ROS param found. Assigning default value:" + self.__position_covariance);

        return True

if __name__ == '__main__':

    rospy.init_node("gps", anonymous=False)

    gps_node = FakeGPSNode()

    if not gps_node.init():
        rospy.logerr("Initialization error.")
        exit()

    loop_rate_hz = 1.0
    loop_rate = rospy.Rate(loop_rate_hz)

    rospy.loginfo("GPS node started.")

    while not rospy.is_shutdown():

        gps_node.update_data(1.0/loop_rate_hz)
        gps_node.publish_data()

        loop_rate.sleep()